Fix EnrichSecurityIT resource cleanup

joegallo opened the pull request:
Closes #84885

So the test failure from https://github.com/elastic/elasticsearch/issues/84885 is really interesting. What happens is that one test randomly fails due to the stacktrace below (which IIUC is an issue that @jbaiera is looking into in the general case), but then that kicks off a comedy of errors where everything fails:

```
org.elasticsearch.xpack.enrich.EnrichSecurityIT > testBasicFlowDate FAILED
    org.elasticsearch.client.WarningFailureException: method [PUT], host [http://127.0.0.1:34431/], URI [/my-index/_doc/1?pipeline=my_pipeline], status line [HTTP/1.1 201 Created]
    Warnings: [index name [.monitoring-es-7-2022.03.10] starts with a dot '.', in the next major version, index names starting with a dot are reserved for hidden indices and system indices]
    {"_index":"my-index","_id":"1","_version":1,"result":"created","_shards":{"total":2,"successful":1,"failed":0},"_seq_no":0,"_primary_term":1}
        at app//org.elasticsearch.client.RestClient.convertResponse(RestClient.java:342)
        at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:312)
        at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:287)
```

See, since that warning is attached to our request, that results in an exception being thrown, so the pipeline that that test just created is never deleted. Which means that the policy referencing that pipeline cannot be deleted and then subsequent tests can't put a new policy because there's already a non-deleted policy with the same name in place.

That is, once any test in this class fails (and so fails to clean up after itself) then all tests in that class from there on out will fail.

This PR fixes that.

joegallo committed:
Fix a typo

joegallo committed:
Add a poison pill

Because of the way this test class is written, as soon as any
assertion fails, then all subsequent tests will fail due to
pipeline/policy conflicts.

This new assertion should only result in a test or two, but because of
the resulting resource issues it causes like six to fail.

joegallo committed:
Clean up pipelines and policies

rather that just policies. This way, if a test accidentally leaves a
pipeline in place (for instance because a test assertion fails) we
don't end up with a cascade of failures where the policy can't be
deleted because the pipeline references it, and then subsequent
policies can't be created because there's already a policy with the
same name, etc.

With this code in place, only the tests that directly violate the
poision pill assertion will fail, and there's no spurious failures or
errors beyond that.

joegallo committed:
Revert "Add a poison pill"

This reverts commit 20cf47a1ad36deeb30bdd3e3f147c20375258293.

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

jbaiera commented on x-pack/plugin/enrich/qa/common/src/main/java/org/elasticsearch/test/enrich/CommonEnrichRestTestCase.java:
```diff
@@ -34,22 +35,44 @@
+            // delete all pipelines
+            List<String> pipelines = new ArrayList<>();
+            try {
+                Map<String, Object> responseMap = toMap(adminClient().performRequest(new Request("GET", "/_ingest/pipeline")));
```
We're not running this in parallel with any other tests that might be creating their own pipelines are we? Is that even something to worry about here?

joegallo replied:
That's a fair point!

My thinking here was that we're already unconditionally deleting all the enrich policies in this cleanup, so we've already got a thread-safety resource-bashing issue. And while that's true, there's probably more tests out there in the world that use ingest pipelines than tests that use enrich policies, so this makes the potential damage bigger than it was before.

I think I can limit this damage pretty well, commit coming.

joegallo replied:
How about 9a67267f414f6023daa4f92ba4c07ce7604ecf2d, @jbaiera?

jbaiera reviewed (approved):
LGTM pending green CI and tiny question

joegallo committed:
Register pipelines for later deletion

Rather that deleting them within the body of the test or deleting all
pipelines unconditionally post test.

joegallo committed:
Merge branch 'master' into fix-enrich-security-it

iverase commented:
This test failed in 8.2 with the same error, are you planing in backporting it? 
https://gradle-enterprise.elastic.co/s/iqehqbkutx3wo/

jbaiera commented:
@iverase This PR doesn't fix that error - it just makes it so that the entire suite doesn't fail. The root issue for that test failure is at https://github.com/elastic/elasticsearch/issues/85506
